// @blibli/blue-dls-utils/src/js/responsive.js
import { reactive } from 'vue'

export const MOBILE_BREAKPOINT = 840

// check whether the device is mobile
export function isMatchedSize (screenSize) {
  const { clientWidth } = window.document.documentElement
  return clientWidth <= screenSize
}

export const config = reactive({
  isMobileDevice: isMatchedSize(MOBILE_BREAKPOINT),
})

export function attachResizeDeviceListener () {
  // listen resize
  window.addEventListener('resize', () => {
    const configArr = Object.keys(config)
    configArr.forEach(size  => {
      const screenSize = size === 'isMobileDevice' ?
        MOBILE_BREAKPOINT : parseInt(size.replace('isDeviceSize',''))
      config[size] = isMatchedSize(screenSize)
    })
  })
}

export function removeResizeDeviceListener (config) {
  // remove resize
  window.addEventListener('resize', () => {
    const configArr = Object.keys(config)
    configArr.forEach(size  => {
      const screenSize = size === 'isMobileDevice' ?
        MOBILE_BREAKPOINT : parseInt(size.replace('isDeviceSize',''))
      config[size] = isMatchedSize(screenSize)
    })
  })
}

export function watchNewDeviceSize (screenSize) {
  config[`isDeviceSize${screenSize}`] = isMatchedSize(screenSize)
}

attachResizeDeviceListener()
