// @blibli/dev-tools/src/utils/config.js
const pathUtil = require('./path')

const fileName = 'blibli-dev-tools.config.js'
const fullPath = pathUtil.getPath(fileName)

const standardFormat = `
module.exports = {
  // attribute to configure packager (Remove this parameter if you don't use packager feature)
  packager: {
    // GROUP_ID (Change this value)
    groupId: 'com.gdn.app-ui',
    // collab project or not
    // isCollab: false,
  },
  // attribute to configure docker (Remove this parameter if you don't use dockerize feature)
  dockerize: {
    // rootDir: 'app',
    // staticDir: '',
  },
  // attribute to configure vite build tools (Remove this parameter if you don't use build tools feature)
  // build: {
  //   vue: 3,
  //   // formats:['cjs', 'es'],
  //   // Source of components(Folder/file name)
  //   sourceComponents: ''
  // },
  // attribute to configure posttest (Remove this parameter if you don't use posttest feature)
  sonar: {
    // Project key (Change this value)
    projectKey: 'com.gdn.app-ui:app',
    // Path lcov.info for adjustement
    // lcovAdjustPaths: ['tests/coverage/lcov.info'],
    // sonar.sources
    // sources: ['.'],
    // sonar.test
    // tests: ['tests'],
    // sonar.exclusions and sonar.coverage.exclusions
    // exclusions: [
    //   // Add files to be excluded in sonar analysis
    // ], 
    // Path lcov to be imported for sonar scanner
    // lcovReportPaths: ['tests/coverage/lcov.info']
  },
  preview: {
    // port: 10001, 
    // publicPath: '', // Only for collab project
    // proxy: {}, // Devs have to pass proxy objects from proxy config
    // replaceVars: {
    //   _STATIC_PUBLIC_PATH: '/',
    //   // _BLIBLI_COLLAB_VER: '1.0.0',
    //   // _STATIC_IMAGE_PUBLIC_PATH: 'aa/',
    // }, // variables to be replaced in production mode
    // noHTMLResponseUrl: [
    //   '/static/',
    //   '/backend/',
    //   // '/s_images/',
    //   // '/serviceworker.js'
    // ]
  }
}`

module.exports = {
  fileName,
  fullPath,
  standardFormat
}
