// @blibli/integration-test-tools/src/utils/string.js
const dot = require('dot-object')

function checkIfStringNotExists (getFileValue, searchedString) {
  const FileValueToString = JSON.stringify(getFileValue)
  const stringExists = FileValueToString.search(searchedString)

  return !(~stringExists)
}

function checkIfObjectNotExits (objPath, obj, searchedValue) {
  const valuePath = dot.pick(objPath, obj)

  const check = !(valuePath !== undefined && valuePath.includes(searchedValue))

  return check
}

function createWhitespace (length) {
  return ' '.repeat(length)
}

function replaceOrInsertNewString (fileValue, oldValue, newValue, doString) {
  const parseStringToArray = fileValue.split('\n')

  parseStringToArray.forEach((item, index) => {
    if (item.includes(oldValue)) {
      const command = {
        replace: () => { parseStringToArray[index] = newValue },
        insert: () => { parseStringToArray.splice(index + 1, 0, newValue) }
      }

      const fn = command[doString]
      fn()
    }
  })

  return parseStringToArray.join('\n')
}

module.exports = {
  checkIfStringNotExists,
  checkIfObjectNotExits,
  createWhitespace,
  replaceOrInsertNewString
}
