// @blibli/dev-tools/resources/build-tools/template.2.cjs.config.js
import { defineConfig } from 'vite'
import { fileURLToPath, URL } from 'node:url'
import { createVuePlugin } from "vite-plugin-vue2"
import path from 'path'
import fs from 'fs'
import resolve from '@rollup/plugin-node-resolve'
import commonjs from '@rollup/plugin-commonjs'
import gzipPlugin from 'rollup-plugin-gzip'
import image from '@rollup/plugin-image'
import terser from '@rollup/plugin-terser'
import cssInjectedByJsPlugin from 'vite-plugin-css-injected-by-js'
import { getBasePath, getBaseDirName, getBaseFileName } from './template.utils'

// css
import url from 'postcss-url'
import babel from '@rollup/plugin-babel'

// Get browserslist config and remove ie from es build targets
const esbrowserslist = fs.readFileSync('./.browserslistrc')
  .toString()
  .split('\n')
  .filter((entry) => entry && entry.substring(0, 2) !== 'ie')

// Extract babel preset-env config, to combine with esbrowserslist
const babelPresetEnvConfig = require('../babel.config')
  .presets.filter((entry) => entry[0] === '@babel/preset-env')[0][1]

const projectRoot = path.resolve(__dirname, '..')
const PATH_NODE_MODULES = path.resolve(projectRoot, 'node_modules')
const fileName = path.basename(__filename);

const external = {}, globals = {}

const baseConfig = {
  plugins: {
    preVue: [],
    replace: {
      'process.env.NODE_ENV': JSON.stringify('production'),
    },
    vue: {
      css: true,
      template: {
        isProduction: true,
      },
      style: {
        postcssPlugins: [
          url({
            url: 'inline'
          })
        ]
      },
    },
    postVue: [
      resolve({
        extensions: ['.js', '.jsx', '.ts', '.tsx', '.vue'],
      }),
      commonjs(),
      image(),
      gzipPlugin()
    ],
    babel: {
      exclude: 'node_modules/**',
      extensions: ['.js', '.jsx', '.ts', '.tsx', '.vue'],
      babelHelpers: 'bundled',
    }
  },
}

// https://vitejs.dev/config/
export default defineConfig({
  plugins: [
    createVuePlugin(),
    cssInjectedByJsPlugin()
  ],
  resolve: {
    alias: [
      {
        find: '@',
        replacement: fileURLToPath(new URL('../src', import.meta.url))
      },
      {
        find: /~(.+)/,
        replacement: fileURLToPath(new URL('../node_modules/$1', import.meta.url))
      }
    ]
  },
  build: {
    emptyOutDir: false,
    lib: {
      entry: getBasePath(fileName, 'cjs'),
	    formats: ['cjs']
    },
    rollupOptions: {
      ...baseConfig,
      external,
      output: {
          dir: getBaseDirName(fileName, 'cjs'),
          entryFileNames: getBaseFileName(fileName, 'cjs', 'min'),
					compact: true,
					format: 'cjs',
					exports: 'auto',
					globals
      },
      plugins: [
        ...baseConfig.plugins.preVue,
        ...baseConfig.plugins.postVue,
        babel(baseConfig.plugins.babel),
        terser({
          output: {
            ecma: 5
          },
        })
      ],
    }
  }
})
